web-services/security/src/main/java/datawave/security/login/DatawavePrincipalLoginModule.java

```diff
@@ -328,6 +351,29 @@ public boolean login() throws LoginException {
                     throw new AccountLockedException(message);
                 }
             }
+
+            /**
+             * Check terminal server to verify that it can directly connect.
+             * This requires the positive check that the terminal server has an approved role
+             * AuthorizedServer or AuthorizedQueryServer.  If TerminalServer does not have the
+             * correct role we will fail the login.  Currently this only checks for UserType.SERVER.
+             * However the predicate could be modified to include a check for UserType.USER.
+             * Logic just streams through the list of ProxiedUsers to get the last element (terminal server).
+             * Make sure it's not null and if it is a server then we want it to have a direct role.
+             */
+
+            DatawavePrincipal principal = (DatawavePrincipal) getIdentity();
+            DatawaveUser terminalServer = principal.getProxiedUsers().stream().reduce((prev, next) -> next).orElse(null);
```

Comment:
This is a pretty novel/clever way to get the last element, but perhaps a call to `Iterables.getLast(...)` might be clearer at a glance.

Reply:
fair point. I can certainly changes this to use Gauva Iterables.  I'm incline to leave it as is at least for now.